### PR TITLE
ethereum-allowlist: Add allowlist pallet for Ethereum contracts

### PR DESCRIPTION
**Pull request by cdamian**
# Description

Adds a new pallet that serves as an allowlist for contracts that can be added by an Ethereum account.

This pallet will be used when processing an EVM `create` call to ensure that no contract is created if the account and the hash of the contract code is not in the allowlist.

# Checklist:

- [x] I have added Rust doc comments to structs, enums, traits and functions
- [x] I have made corresponding changes to the documentation
- [x] I have performed a self-review of my code
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by mustermeiszer — COMMENTED** [review]
Looks good! Just minor improvement regarding storage and logic to allow multiple calls of same hash. 

I would impl those 
* https://github.com/paritytech/frontier/blob/master/frame/ethereum/src/lib.rs#L108-L167
methods on the pallet level as a proxy and check when it is create if somebody is allowed to call it. 

Concern: 
* How do we remove the allowance once the extrinsic is through? I currently think the only way is to also proxy `pallet_ethereum::Call::Transact{..}` and force users to use this one as a self-contained call when they want to create stuff. 

Not sure how this works with the EVM side though ...

**Comment by mustermeiszer** [comment]
cc @branan regarding proxying this

File: pallets/ethereum-allowlist/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	/// Storage for the hashes of the Ethereum contract byte codes that an
+	/// Ethereum account is allowed to create.
+	#[pallet::storage]
+	#[pallet::getter(fn account_code_hashes)]
+	pub type AccountCodeHashes<T: Config> =
+		StorageDoubleMap<_, Blake2_128Concat, H160, Blake2_128Concat, H256, (), ValueQuery>;
```
[the remark this answers is not in the document]
Comment (cdamian) [reply pallets/ethereum-allowlist/src/lib.rs:54]: Hmm, not sure how that would work. I thought this is just an allowlist for a contract that can be created on our EVM. Do we want to keep track of how many times a contract with the same code is created or what would be the purpose of that?

File: runtime/development/Cargo.toml [hunk 2]
```diff
@@ -124,6 +124,7 @@ pallet-connectors-gateway = { path = "../../pallets/connectors-gateway", default
 pallet-crowdloan-claim = { path = "../../pallets/crowdloan-claim", default-features = false }
 pallet-crowdloan-reward = { path = "../../pallets/crowdloan-reward", default-features = false }
 pallet-data-collector = { path = "../../pallets/data-collector", default-features = false }
+pallet-ethereum-allowlist = { path = "../../pallets/ethereum-allowlist", default-features = false }
```
Comment (lemunozm) [inline runtime/development/Cargo.toml:127]: Here is the issue I think, you need to add this pallet to std/runtime-benchmarks/try-runtime

Comment (cdamian) [reply runtime/development/Cargo.toml:127]: Yup, I already have those changes locally and it's fine, I'm tackling the clippy issue now.

**Comment by mustermeiszer** [comment]
closing as we currently do not need that.

**Pull request closed without merging**